server/controllers/finance/cash.js

```diff
@@ -0,0 +1,247 @@
+
+  db.exec(sql)
+  .then(function (rows) {
+    res.status(200).json(rows);
+  })
+  .catch(next)
+  .done();
+};
+
+/**
+ * GET /cash/:uuid
+ *
+ * Get the details of a particular cash payment.  This endpoint will return a
+ * record in the following format:
+ * {
+ *   uuid : "..",
+ *   items : [ {}, {} ]    // items hitting invoices, if applicable
+ *   is_caution : true/false,
+ *   canceled: true/false, // indicates if a cash_discard record exists
+ *   ...
+ *  }
+ */
+exports.details = function details(req, res, next) {
+  'use strict';
+
+  var uuid = req.params.uuid;
+
+  lookupCashRecord(uuid, req.codes)
+  .then(function (record) {
+    res.status(200).json(record);
+  })
+  .catch(next)
+  .done();
+};
+
+
+/**
+ * POST /cash
+ * Creates a cash payment against several sales or a cautionary payment.  The
+ * API also supports future offline functionality by either accepting a UUID or
+ * generating it if it is not present.
+ */
+exports.create = function create(req, res, next) {
+  'use strict';
+
+  var data = req.body.payment;
+
+  // generate a UUID if it not provided.
+  // @TODO - should we validate that this is an _actual_ uuid, or should this
+  // be sufficient?
+  data.uuid = data.uuid || uuid();
+
+  // account for the cash items
+  var items =  data.items;
+
+  // remove the cash items so that the SQL query is properly formatted
+  delete data.items;
+
+  // if items exist, tranform them into an array of arrays for db formatting
+  if (items) {
+    items = items.map(function (item) {
+      item.uuid = item.uuid || uuid();
+      item.cash_uuid = data.uuid;
+      return [
+        item.uuid, item.cash_uuid,
+        item.amount, item.invoice_uuid
+      ];
+    });
+  }
+
+  var writeCashSql =
+    'INSERT INTO cash SET ?;';
+
+  var writeCashItemsSql =
+    'INSERT INTO cash_item (uuid, cash_uuid, amount, invoice_uuid) ' +
+    'VALUES (?);';
+
+  var transaction = db.transaction();
+  transaction.addQuery(writeCashSql, [ data ]);
+
+  // only add the "items" query if we are NOT making a caution
+  // cautions do not have items
+  if (!data.is_caution) {
+    transaction.addQuery(writeCashItemsSql, items);
+  }
+
+  transaction.execute()
+  .then(function () {
+    res.status(201).json({ uuid : data.uuid });
+  })
+  .catch(next)
+  .done();
+};
+
+
+/**
+ * PUT /cash/:uuid
+ * Updates the non-financial details associated with a cash payment.
+ * NOTE - this will not update the cash_item or cash_discard tables.
+ */
+exports.update = function update(req, res, next) {
+  'use strict';
+
+  var uuid = req.params.uuid;
+  var updateSql = 'UPDATE cash SET ?;';
+
+  // protected database fields that are unavailable for updates.
+  var protect = [
+    'is_caution', 'amount', 'user_id', 'cashbox_id',
+    'currency_id', 'date', 'uuid', 'project_id'
+  ];
+
+  // loop through update keys and ensure that we are only updating non-protected
+  // fields
+  var keys = Object.keys(req.body);
```

Comment:
[According the Open Web Application Security Project](https://www.owasp.org/index.php/REST_Security_Cheat_Sheet),
> A data contextual check needs to be done, server side, with each request.

From an [HTML5 Rocks article](http://www.html5rocks.com/en/tutorials/forms/constraintvalidation/) on client-side validation:
>It's important to remember that even with these new APIs client side validation does not remove the need for server side validation. Malicious users can easily workaround any client side constraints, and, HTTP requests don't have to originate from a browser.

> Therefore client side validation should always be treated as a progressive enhancement to the user experience; all forms should be usable even if client side validation is not present.

> Since you have to have server side validation anyways, if you simply have your server side code return reasonable error messages and display them to the end user you have a built in fallback for browsers that don't support any form of client side validation.